hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -174,68 +151,67 @@ abstract class MatrixHybridReader extends TableReader with MatrixReader {
 
-    MatrixNativeReader(path, options, rowsSpec, entriesSpec, colsSpec, spec)
+    val intervals = params.options.map(_.intervals)
+    if (intervals.nonEmpty && !spec.indexed)
+      fatal("""`intervals` specified on an unindexed matrix table.
+              |This matrix table was written using an older version of hail
+              |rewrite the matrix in order to create an index to proceed""".stripMargin)
+
+    new MatrixNativeReader(params, spec)
   }
 
-  def fromJson(fs: FS, readerJV: JObject): MatrixNativeReader = {
-    val path = readerJV \ "path" match {
+  def fromJValue(fs: FS, jv: JValue): MatrixNativeReader = {
+    val path = jv \ "path" match {
       case JString(s) => s
     }
 
-    val options = readerJV \ "options" match {
+    val options = jv \ "options" match {
       case optionsJV: JObject =>
-        Some(NativeReaderOptions.fromJson(optionsJV))
+        Some(NativeReaderOptions.fromJValue(optionsJV))
       case JNothing => None
     }
 
-    read(fs, path, options)
+    MatrixNativeReader(fs, MatrixNativeReaderParameters(path, options))
   }
 }
 
-case class MatrixNativeReader(
+case class MatrixNativeReaderParameters(
   path: String,
-  options: Option[NativeReaderOptions],
-  rowsSpec: AbstractTableSpec,
-  entriesSpec: AbstractTableSpec,
-  colsSpec: AbstractTableSpec,
+  options: Option[NativeReaderOptions])
+
+class MatrixNativeReader(
+  val params: MatrixNativeReaderParameters,
   spec: AbstractMatrixTableSpec
 ) extends MatrixReader {
+  def pathsUsed: Seq[String] = FastSeq(params.path)
 
-  def pathsUsed: Seq[String] = FastSeq(path)
-  lazy val columnCount: Option[Int] = Some(colsSpec
+  lazy val columnCount: Option[Int] = Some(spec.colsSpec
     .partitionCounts
     .sum
     .toInt)
 
-  def partitionCounts: Option[IndexedSeq[Long]] = if (intervals.isEmpty) Some(spec.partitionCounts) else None
+  def partitionCounts: Option[IndexedSeq[Long]] = if (params.options.isEmpty) Some(spec.partitionCounts) else None
 
   def fullMatrixType: MatrixType = spec.matrix_type
 
-  private def intervals = options.map(_.intervals)
-
-  if (intervals.nonEmpty && !spec.indexed(path))
-    fatal("""`intervals` specified on an unindexed matrix table.
-            |This matrix table was written using an older version of hail
-            |rewrite the matrix in order to create an index to proceed""".stripMargin)
-
   override def lower(mr: MatrixRead): TableIR = {
-    val rowsPath = path + "/rows"
-    val entriesPath = path + "/entries"
-    val colsPath = path + "/cols"
+    val rowsPath = params.path + "/rows"
+    val entriesPath = params.path + "/entries"
+    val colsPath = params.path + "/cols"
 
     if (mr.dropCols) {
       val tt = TableType(mr.typ.rowType, mr.typ.rowKey, mr.typ.globalType)
-      val trdr: TableReader = TableNativeReader(rowsPath, options, rowsSpec)
+      val trdr: TableReader = new TableNativeReader(TableNativeReaderParameters(rowsPath, params.options), spec.rowsSpec)
```

Comment:
are we missing a PR? I see TableNativeReader as a case class and no TableNativeReaderParameters

Reply:
Never mind, GitHub had collapsed it from search, and I missed it among other files